app/controllers/issues_controller.rb

```diff
@@ -0,0 +1,101 @@
+# frozen_string_literal: true
+
+class IssuesController < ApplicationController
+  load_and_authorize_resource :project
+  load_and_authorize_resource through: :project
+  before_action :load_valid_assignees, only: %i[new edit update create]
+
+  # GET /projects/:project_id/issues/:id
+  def show
+    @comment = Comment.new
+    @pagy, @comments = pagy(Comment.where(commentable: @issue))
+    respond_to do |format|
+      format.html
+    end
+  end
+
+  # GET /projects/:project_id/issues/new
+  def new
+    respond_to do |format|
+      format.html
+    end
+  end
+
+  # POST /projects/:project_id/issues
+  def create
+    # Here the company is being set via the cancancan ability written as,
+    # can :create, Issue, company_id: user.company_id
+
+    # Similar is the case for user
+    # can :create, Issue, creator_id: user.id
+
+    # Project for the issue is also set automatically here via the cancancan load_and_authorize statements
+    # as given above
+
+    respond_to do |format|
+      if @issue.save
+        format.html { redirect_to project_issue_path(@project, @issue) }
+      else
+        format.html { render 'new' }
+      end
+    end
+  end
+
+  # GET /projects/:project_id/issues/:id/edit
+  def edit
+    respond_to do |format|
+      format.html
+    end
+  end
+
+  # PATCH /projects/:project_id/issues/:id(.:format)
+  # PUT /projects/:project_id/issues/:id(.:format)
+  def update
+    respond_to do |format|
+      if @issue.update(issue_params)
+        format.html { redirect_to project_issue_path(@project, @issue) }
+      else
+        format.html { render 'edit' }
+      end
+    end
+  end
+
+  # DELETE /projects/:project_id/issues/:id
+  def destroy
+    respond_to do |format|
+      if @issue.destroy
+        format.html { redirect_to project_path(params[:project_id]) }
+      else
+        format.html { redirect_back fallback_location: root_path }
+      end
+    end
+  end
+
+  # PATCH /projects/:project_id/issues/:id/update_status
+  def update_status
+    status_str = params[:status]
```

Comment:
first check if params[:status] being sent is a valid event name, maybe something like `events_constant.include?(params[:status)`, if that is true only then call line-77

Reply:
It is done below we are checking if the event name is valid in below statement where find is called. If the name is invalid it will just flash the incorrect event message

Reply:
what if someone deliberately sends a garabage params[:status] by inspecting and editing the fields in the browser? Do we want `update_event = @issue.aasm.events(permitted: true).find { |event| event.name.to_s.humanize == status_str }` these to computations to run for garbage input as well? No. Which is why filter the status being sent first, and then give to further computations. `%w[valid_status1 valid_status2..].includes?(params[:status])`

Reply:
It will work almost like this. Even if garbage data is in the params. If the data is garbage or like unknown event the find method would not be able to find the event and nil will be stored in update_event and the @issue.public_send will not be called and instead the flash with error will be called

Reply:
my point here was if params has garbage data, we don't need to hit memory, we don't need to call `@issue.aasm.events(permitted: true)` and load all the events into memory, we don't need to iterate uselessly over the loaded events, that's computation going to waste. Simply filter first if params[:status] has a value which is a valid event name before we get into computations.

Reply:
done